src/api.jl

```diff
@@ -29,35 +29,47 @@ function seed(v)
     return _seed.(Ref(vv), CartesianIndices((ax, ax)))
 end
 
-function D(f)
-    # grad
-    function deriv(arg::AbstractArray)
-        # always chunk
-        arg_partial = seed(arg)
-        darr = dualrun(()->DualArray(arg, arg_partial))
-        res = dualrun(()->f(darr))
-        diffres = extract_diffresult(allpartials(res))
-        return diffres
-    end
-    # scalar
-    function deriv(x)
-        dx = one(x)
-        res = dualrun() do
-            dualized = Dual(x, dx)
-            f(dualized)
-        end
-        return map(partials, res)
-    end
-    return deriv
+###
+### Derivative object:
+###
+
+"""
+    D(f, x)
+
+`D(f, x) * v` computes ``df/dx * v``
+"""
+struct D{T,F}
+    f::F
+    x::T
 end
 
-#=
-# scalar case: f: R -> something
-D(sin)(1.0)
-D(x->[x, x^2])(3)
+# WARNING: It assume that the number type is commutative
+Base.:*(v, dd::D{<:Number}) = dd * v
+function Base.:*(dd::D{<:Number}, v)
+    is_I = v isa UniformScaling
+    derivative_dual = dualrun() do
+        dual = Dual(dd.x, is_I ? v.λ : v)
+        dd.f(dual)
+    end
+    derivative = map(partials, derivative_dual)
+    is_I && (derivative *= I)
+    return derivative
+end
 
-# gradient case: f: R^n -> R
-D(sum)([1,2,3])
+function Base.:*(dd::D{<:AbstractArray}, II::UniformScaling)
+    # always chunk
```

Comment:
why?

Reply:
I didn't implement the other stuff yet. Always doing vector mode is easy to write, so I wrote that first.

Reply:
oh wait I read it wrong. Yeah, always chunk makes sense.